onnx/test/version_converter/automatic_downgrade_test.py

```diff
@@ -0,0 +1,52 @@
+# Copyright (c) ONNX Project Contributors
+
+# SPDX-License-Identifier: Apache-2.0
+import unittest
+
+import automatic_conversion_test_base
+import numpy as np
+import parameterized
+
+import onnx
+from onnx import helper
+
+#####################################################################################
+# Every test calls _test_op_conversion to downgrade a model from the most recent opset version
+# to a early version and runs checker + shape inference on the downgraded model.
+####################################################################################
+
+
+class TestAutomaticDowngrade(automatic_conversion_test_base.TestAutomaticConversion):
+    def _test_op_downgrade(self, op: str, *args, **kwargs):
+        self._test_op_conversion(op, *args, **kwargs, is_upgrade=False)
+
+    @parameterized.parameterized.expand(
+        [
+            "ReduceL1",
+            "ReduceL2",
+            "ReduceLogSum",
+            "ReduceLogSumExp",
+            "ReduceMean",
+            "ReduceMax",
+            "ReduceMin",
+            "ReduceProd",
+            "ReduceSum",
+            "ReduceSumSquare",
+        ]
+    )
+    def test_reduce_ops(self, op) -> None:
```

Comment:
I wonder if we can just use the backend test to make a pass on all operators.

Reply:
as you can see from upgrade test where the test input, output, and attributes are specific to individual operators. It is difficult to iterate over all ops without specify input, output and attribute specifics of the operator. it is true that backend test has input and output and may be possible to do the test for upgrade. however, it is not necessary that all backend tests will pass upgrade tests (as an example, the test case may not be valid for the from_version of the op). Further with downgrade test, because it is optional, downgrade path for many ops does not work.

Reply:
We may be able to catch bugs while doing that. A test valid for a specific opset should be the same for a newer opset. I think that's what is expected when a model is upgraded unless the operator was deprecated. I'll close this comment. Maybe we can reopen the discussion if an issue is raised about this functionality.

Reply:
I agree we should have a more comprehensive set of test cases applicable for different uses